std-bits/table/src/main/java/org/enso/table/data/column/operation/BinaryOperationBoolean.java

```diff
@@ -0,0 +1,105 @@
+  private final boolean preserveNulls;
+
+  public BinaryOperationBoolean() {
+    this(true);
+  }
+
+  public BinaryOperationBoolean(boolean preserveNulls) {
+    super(BooleanType.INSTANCE);
+    this.preserveNulls = preserveNulls;
+  }
+
+  @Override
+  public ColumnStorage<Boolean> applyMap(ColumnStorage<?> left, Object rightValue) {
+    assert canApplyMap(left, rightValue);
+
+    if (preserveNulls && rightValue == null) {
+      // Return an all null column
+      return BoolStorage.makeEmpty(left.getSize());
+    }
+
+    if (rightValue != null && !(rightValue instanceof Boolean)) {
+      throw new IllegalArgumentException(
+          "Unsupported right value type " + rightValue.getClass() + ".");
+    }
+
+    boolean rightIsNothing = rightValue == null;
+    boolean rightBoolean = !rightIsNothing && (boolean) rightValue;
+
+    if (left instanceof BoolStorage leftBoolStorage) {
+      var result = applyMapOverBoolStorage(leftBoolStorage, rightBoolean, rightIsNothing);
+      if (result != null) {
+        return result;
+      }
+    }
+
+    return StorageIterators.buildOverBooleanStorage(
+        BooleanType.INSTANCE.asTypedStorage(left),
+        preserveNulls,
+        makeStorageBuilder(left.getSize(), left.getType(), BooleanType.INSTANCE),
+        (b, index, value, isNothing) -> {
+          Boolean result = applySingle(value, isNothing, rightBoolean, rightIsNothing);
+          if (result == null) {
+            b.appendNulls(1);
+          } else {
+            b.appendBoolean(result);
+          }
+        });
+  }
+
+  @Override
+  public ColumnStorage<Boolean> applyZip(ColumnStorage<?> left, ColumnStorage<?> right) {
+    assert canApplyZip(left, right);
+
+    if (right.getType() instanceof NullType) {
+      return applyMap(left, null);
+    }
+
+    if ((left instanceof BoolStorage leftBoolStorage)
+        && (right instanceof BoolStorage rightBoolStorage)) {
+      var result = applyZipOverBoolStorage(leftBoolStorage, rightBoolStorage);
+      if (result != null) {
+        return result;
+      }
+    }
+
+    return StorageIterators.zipOverBooleanStorages(
+        BooleanType.INSTANCE.asTypedStorage(left),
+        BooleanType.INSTANCE.asTypedStorage(right),
+        s -> makeStorageBuilder(s, left.getType(), right.getType()),
+        preserveNulls,
+        (index, value, isNothing, rightValue, rightIsNothing) ->
+            applySingle(value, isNothing, rightValue, rightIsNothing));
+  }
+
+  @Override
+  protected BuilderForBoolean makeStorageBuilder(
+      long size, StorageType<?> leftType, StorageType<?> rightType) {
+    return BooleanType.INSTANCE.makeBuilder(size, BlackholeProblemAggregator.INSTANCE);
+  }
+
+  protected ColumnBooleanStorage applyMapOverBoolStorage(
+      BoolStorage left, boolean rightBoolean, boolean rightIsNothing) {
+    return null;
+  }
+
+  protected ColumnBooleanStorage applyZipOverBoolStorage(BoolStorage left, BoolStorage right) {
+    return null;
+  }
```

Comment:
I think it would be worth adding a comment that these should be overridden to return a non-null value if a more efficient 'specialization' should be applied in the given operation / set of arguments.

Without a comment it's not that trivial for future implementors of this class to figure out what this method is and when they should override it.

I'd also suggest to indicate this partially in the name (e.g. `applySpecializedMapOverBoolStorage`) but still a comment could be worth it.

Ideally maybe also a mention in the class header.